### DISCUSSION TITLE
RP2040 communication with MAX and processing ide

### DISCUSSION THREAD

**Discussion by MishalJasmine** [micropython/micropython#11457] [RP2040 / Pico] [answered]
In my project, RP2040 is connected to MAX3001 with which it communicates using SPI. The values are read back from MAX and plotted in a open source software called [processing](https://processing.org/). I have written the scripts in Micropython. But the issue is with the ports. 
Data is sent to the processing ide using UART. If I connect the port (COM4) in the processing ide, its gives no port available error since the same port (COM4)  is already connected in Thonny. 

Please let me know how to resolve this.

**Comment by Josverl** [comment]
The rp2040 has two uarts, so the simplest software solution would be to add a uart to usb convertor, and thus have access to two com ports, one for Thonny, and the other for your datastream.
[One of many options](https://www.aliexpress.com/item/CP2102-MICRO-USB-to-UART-TTL-Module-6Pin-Serial-Converter-UART-STC-Replace-FT232/32856532426.html) 

On windows opened Com ports cannot be shared between processes, so Thonny and Processing cannot both acces COM4:.

On Linux sharing ports is possible, but will require Thonny, processing, and your Micropython code to accommodate for that

**Comment by MishalJasmine** [comment]
unfortunately I am using windows. Is there any solution for live streaming of data from MAX?

**Comment by robert-hh** [comment, marked as answer]
> But the issue is with the ports. Data is sent to the processing ide using UART. If I connect the port (COM4) in the processing ide, its gives no port available error since the same port (COM4) is already connected in Thonny.

The RP2040 port does usually use the USB interface for communication, which is mapped into the PC's OS asl COM port. Maybe COM4 in your case. You can use the hardware UART devices in addition, but then you have to attach e.g. a UART/USB adapter. Other option:

- Disconnect Thonny and use the USB-UART for communication. On the board rp2040, use print() instead of uart.write(). Then it should be possible to connect the processing IDE.

**Reply by MishalJasmine** [reply]
After uploading the code to the board rp2040 and running it, disconnected from the thonny and then tried connecting in processing. It worked.
